algorithms/src/snark/marlin/ahp/prover/round_functions/fourth.rs

```diff
@@ -22,37 +22,35 @@ use crate::{
         ahp::{verifier, AHPError, AHPForR1CS},
         prover,
         MarlinMode,
-    },
+    }, fft::DensePolynomial,
 };
 
+use itertools::Itertools;
 use rand_core::RngCore;
 use snarkvm_fields::PrimeField;
 
 impl<F: PrimeField, MM: MarlinMode> AHPForR1CS<F, MM> {
     /// Output the fourth round message and the next state.
-    pub fn prover_fourth_round<R: RngCore>(
+    pub fn prover_fourth_round<'a, R: RngCore>(
         verifier_message: &verifier::ThirdMessage<F>,
-        state: prover::State<F, MM>,
+        mut state: prover::State<'a, F, MM>,
         _r: &mut R,
     ) -> Result<prover::FourthOracles<F>, AHPError> {
-        let verifier::ThirdMessage { r_b, r_c, .. } = verifier_message;
-        let [mut lhs_a, mut lhs_b, mut lhs_c] = state.lhs_polynomials.unwrap();
-        lhs_b *= *r_b;
-        lhs_c *= *r_c;
-
-        lhs_a += &lhs_b;
-        lhs_a += &lhs_c;
-        let h_2 = LabeledPolynomial::new("h_2".into(), lhs_a, None, None);
+        let verifier::ThirdMessage { r_a, r_b, r_c } = verifier_message;
```

Comment:
Let's add an `iter()` method to `verifier::ThirdMessage` which outputs tuples of the form `(r_a[i], r_b[i], r_c[i])`

Reply:
Done, looks and feels cleaner indeed, though the iterators do add net 16 lines of code